Rubberduck.Parsing/VBA/ProjectManager.cs

```diff
@@ -1,54 +1,41 @@
-﻿using System;
-using System.Collections.Generic;
+﻿using System.Collections.Generic;
 using System.Linq;
 using Rubberduck.VBEditor;
 using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
 
 namespace Rubberduck.Parsing.VBA
 {
-    public class ProjectManager : IProjectManager
+    public class ProjectManager : ProjectManagerBase
     {
-        private readonly RubberduckParserState _state;
-        private readonly IVBE _vbe;
+        private const int _maxDegreeOfQMNCreationParallelism = -1;
 
         public ProjectManager(
             RubberduckParserState state,
             IVBE vbe)
-        {
-            if (state == null)
-            {
-                throw new ArgumentNullException(nameof(state));
-            }
-            if (vbe == null)
-            {
-                throw new ArgumentNullException(nameof(vbe));
-            }
-
-            _state = state;
-            _vbe = vbe;
-        }
+        :base(state, 
+            vbe)
+        { }
 
 
-        public IReadOnlyCollection<IVBProject> Projects
+        public override IReadOnlyCollection<QualifiedModuleName> AllModules()
         {
-            get
-            {
-                return _state.Projects.AsReadOnly();
-            }
-        }
+            var components = Projects.SelectMany(project => project.VBComponents);
 
-        public IReadOnlyCollection<QualifiedModuleName> AllModules()
-        {
-            return Projects.SelectMany(project => project.VBComponents)
-                            .Select(component => new QualifiedModuleName(component))
-                            .ToHashSet()
-                            .AsReadOnly(); ;
-        }
+            var options = new ParallelOptions();
+            options.MaxDegreeOfParallelism = _maxDegreeOfQMNCreationParallelism;
 
+            var modules = new ConcurrentBag<QualifiedModuleName>();
+            Parallel.ForEach(components,
+                options,
+                component =>
+                {
+                    modules.Add(new QualifiedModuleName(component));
```

Comment:
So.... COM STA isn't blowing a gasket over this? Does the caller handle any `COMException` by falling back to the synchronous code?